diff --git a/DeviceLockInput.qml b/DeviceLockInput.qml
index cf8fe14..e4a421f 100644
--- a/usr/lib/qt5/qml/com/jolla/settings/system/DeviceLockInput.qml
+++ b/usr/lib/qt5/qml/com/jolla/settings/system/DeviceLockInput.qml
@@ -3,7 +3,7 @@ import Sailfish.Silica 1.0
 import com.jolla.settings.system 1.0
 import org.nemomobile.devicelock 1.0
 
-PinInput {
+MazeInput {
     id: input
 
     property bool requestSignalOnly: true
--- a/usr/lib/qt5/qml/com/jolla/settings/system/qmldir
+++ b/usr/lib/qt5/qml/com/jolla/settings/system/qmldir
@@ -9,6 +9,7 @@ DeviceLockQuery 1.0 DeviceLockQuery.qml
 LanguagePickerPage 1.0 LanguagePickerPage.qml
 NetworkAddressField 1.0 NetworkAddressField.qml
 PinInput 1.0 PinInput.qml
+MazeInput 1.0 MazeInput.qml
 SimActivationPullDownMenu 1.0 SimActivationPullDownMenu.qml
 SimLockedNotice 1.0 SimLockedNotice.qml
 SimPinInput 1.0 SimPinInput.qml
