commit 7b6ce7d0daaf286685e370e4c1c6df7b73f1137a
Author: CODeRUS <coderusinbox@gmail.com>
Date:   Wed Dec 17 01:24:02 2014 +0500

    MazeLock 0.0.3

diff --git a/usr/share/lipstick-jolla-home-qt5/main.qml b/usr/share/lipstick-jolla-home-qt5/main.qml
index 6c4fa88..0d59e00 100644
--- a/usr/share/lipstick-jolla-home-qt5/main.qml
+++ b/usr/share/lipstick-jolla-home-qt5/main.qml
@@ -273,7 +273,7 @@ ApplicationWindow {
             maximumFlickVelocity: 4000*Theme.pixelRatio
             highlightMoveDuration: 300
             pressDelay: 0
-            interactive: !closeApplicationEnabled && !removeApplicationEnabled && !launcher.openedChildFolder && !desktop.animating
+            interactive: !closeApplicationEnabled && !removeApplicationEnabled && !launcher.openedChildFolder && !desktop.animating && !(deviceLockUI.item && deviceLockUI.item.busy)
             quickScroll: false
 
             onMovingChanged: {
diff --git a/usr/share/lipstick-jolla-home-qt5/main/DeviceLockView.qml b/usr/share/lipstick-jolla-home-qt5/main/DeviceLockView.qml
index 4593c79..ab7ef3e 100644
--- a/usr/share/lipstick-jolla-home-qt5/main/DeviceLockView.qml
+++ b/usr/share/lipstick-jolla-home-qt5/main/DeviceLockView.qml
@@ -13,6 +13,8 @@ import org.nemomobile.ngf 1.0
 
 DeviceLockInput {
     id: pininput
+    property int attempt: attemptCount.value
+    property bool busy: false
 
     function reset() {
         pininput._displayedPin=""
diff --git a/usr/lib/qt5/qml/com/jolla/settings/system/PinInput.qml b/usr/lib/qt5/qml/com/jolla/settings/system/PinInput.qml
index 7c1d86c..3e66bdf 100644
--- a/usr/lib/qt5/qml/com/jolla/settings/system/PinInput.qml
+++ b/usr/lib/qt5/qml/com/jolla/settings/system/PinInput.qml
@@ -2,12 +2,41 @@ import QtQuick 2.0
 import Sailfish.Silica 1.0
 import MeeGo.QOfono 0.2
 import org.nemomobile.lipstick 0.1
+import org.nemomobile.configuration 1.0
 
 Item {
     id: root
+    // new properties
+    property bool useMaze: !root.hasOwnProperty("simManager")
+    onUseMazeChanged: {
+        if (!_showDigitPad && !useMaze) {
+            alphanumProxy.forceActiveFocus()
+        }
+    }
+    property int captchaLine: 1
+    property bool captcha: false
+    onCaptchaChanged: {
+        if (captcha) {
+            captchaLine = Math.floor((Math.random() * mazeLock.pinx) + 1)
+        }
+    }
+    Connections {
+       target: root
+       ignoreUnknownSignals: true
+       onAttemptChanged: {
+           if (attempt > 2 && parseInt(attempt % 3) == 0) {
+               root.captcha = true
+           }
+       }
+    }
 
     // read-only
     property string enteredPin
+    onEnteredPinChanged: {
+        if (mazeLock.visible) {
+            mazeLock.repaint()
+        }
+    }
     property bool emergency
     property bool enteringNewPin
 
@@ -71,7 +100,7 @@ Item {
     property QtObject _voiceCallManager
 
     property bool showDigitPad: true
-    property bool _showDigitPad: showDigitPad || emergency
+    property bool _showDigitPad: emergency || (!useMaze && showDigitPad)
     property bool visibleInDashboard
 
     signal pinConfirmed()
@@ -101,7 +130,12 @@ Item {
     }
 
     function _clickedConfirmButton() {
-        if (enteringNewPin) {
+        if (!useMaze && ((keypad.visible && enteredPin == "62935625") || (alphanumProxy.visible && enteredPin == "mazelock"))) {
+           useMaze = true
+           enteredPin = ""
+           _displayedPin = ""
+        }
+        else if (enteringNewPin) {
             if (_newPin === "") {
                 _pinConfirmTitleText = confirmNewPinText
                 _badPinWarning = ""
@@ -121,7 +155,42 @@ Item {
                 }
             }
         } else {
-            pinConfirmed()
+            if (captcha) {
+                if (useMaze) {
+                    var startPos = mazeLock.pinx * (captchaLine - 1) + 1
+                    var captchaString = mazeLock.chars.substr(startPos, mazeLock.pinx)
+                    if (enteredPin == captchaString) {
+                        captcha = false
+                    }
+                    else {
+                        var oldLine = captchaLine
+                        while (captchaLine == oldLine) {
+                            captchaLine = Math.floor((Math.random() * mazeLock.pinx) + 1)
+                        }
+                    }
+                }
+                else {
+                    if (keypad.visible) {
+                        if (enteredPin == "13795") {
+                            captcha = false
+                        }
+                    }
+                    else {
+                        if (enteredPin == "jolla") {
+                            captcha = false
+                        }
+                    }
+                }
+                
+                if (captcha && _feedbackEffect) {
+                    _feedbackEffect.play()
+                }
+                enteredPin = ""
+                _displayedPin = ""
+            }
+            else {
+                pinConfirmed()   
+            }
         }
     }
 
@@ -139,9 +208,14 @@ Item {
                 return
             }
             obfuscateLastDigit.stop()
-            _displayedPin = _passwordString(_displayedPin.length) + digit
+            if (useMaze && enteringNewPin) {
+               _displayedPin += digit
+            }
+            else {
+               _displayedPin = _passwordString(_displayedPin.length) + (useMaze && mazeLockSettings.maskImmediately ? _passwordCharacter : digit)
+               obfuscateLastDigit.start()
+           }
             enteredPin += digit
-            obfuscateLastDigit.start()
             if (minimumLength > enteredPin.length) {
                 _overridingWarningText = pinShortLengthWarning
             } else if (minimumLength === enteredPin.length) {
@@ -245,7 +319,8 @@ Item {
         color: root.warningTextColor
 
         font.pixelSize: Theme.fontSizeSmall
-        text: root._overridingWarningText !== "" ? root._overridingWarningText : root.warningText
+        text: root.captcha ? (useMaze ? qsTr("Draw line number %1 to continue").arg(captchaLine) : (keypad.visible ? "Enter \"13795\" to continue" : "Enter \"jolla\" to continue"))
+                           : root._overridingWarningText !== "" ? root._overridingWarningText : root.warningText
     }
 
     BackgroundItem {
@@ -299,7 +374,7 @@ Item {
             leftMargin: 2* Theme.paddingSmall
             right: backspace.left
             rightMargin: Theme.paddingSmall
-            bottom: showEmergencyButton || keypad.visible ? keypad.top : alphanumProxy.bottom
+            bottom: showEmergencyButton || keypad.visible ? keypad.top : (useMaze ? mazeLock.top : alphanumProxy.bottom)
             bottomMargin: Theme.itemSizeSmall / 2
         }
         horizontalAlignment: Text.AlignRight
@@ -378,6 +453,215 @@ Item {
         onPressed: root._pushPinDigit(number + "")
     }
 
+    // mazelock configuration settings. loaded once after lipstick start.
+    ConfigurationGroup {
+        id: mazeLockSettings
+        path: "/desktop/nemo/devicelock/mazelock"
+        property int size: 4
+        property bool colored: true
+        property bool maskImmediately: true
+    }
+
+    Item {
+        id: mazeLock
+        width: parent.width
+        height: width
+        visible: !emergency && useMaze
+        anchors.bottom: parent.bottom
+
+        property int pinx: showDigitPad ? 3 : (mazeLockSettings.size < 3 ? 3 : (mazeLockSettings.size > 6 ? 6 : mazeLockSettings.size))
+        property int piny: pinx
+
+        property var colors: ["#FFFFFF", "#FFFF80", "#FFFF00", "#FF80FF",
+                              "#FF8080", "#FF8000", "#FF00FF", "#FF0080",
+                              "#FF0000", "#80FFFF", "#80FF80", "#80FF00",
+                              "#8080FF", "#808080", "#808000", "#8000FF",
+                              "#800080", "#800000", "#00FFFF", "#00FF80",
+                              "#00FF00", "#0080FF", "#008080", "#008000",
+                              "#0000FF", "#000080", "#000000"]
+        property string chars: "#1234567890abcdefghijklmnopqrstuvwxyz"
+        //123
+        //456
+        //789
+
+        //1234
+        //5678
+        //90ab
+        //cdef
+
+        //12345
+        //67890
+        //abcde
+        //fghij
+        //klmno
+
+        //123456
+        //7890ab
+        //cdefgh
+        //ijklmn
+        //opqrst
+        //uvwxyz
+
+        property var nodes: [] // unused?
+
+        function repaint() {
+            canvas.requestPaint()
+        }
+
+        MouseArea {
+            id: pinArea
+            width: parent.width
+            height: width
+            anchors.centerIn: parent
+            onPressedChanged: if (root.hasOwnProperty("busy")) root.busy = pressed
+
+            Repeater {
+               id: mazeCreator
+               model: mazeLock.pinx * mazeLock.piny
+               delegate: pinComponent
+            }
+
+            Canvas {
+                id: canvas
+                anchors.fill: parent
+                z: 0
+
+                onPaint: {
+                    var ctx = getContext("2d")
+                    ctx.clearRect(0, 0, width, height)
+                    ctx.lineWidth = 4
+                    ctx.beginPath()
+                    if (root.enteredPin.length > 1) {
+                        for (var i = 0; i < root.enteredPin.length; i++) {
+                            var node = pinArea.getNode(mazeLock.chars.indexOf(root.enteredPin[i]))
+                            var point = pinArea.getNodePoint(node.x, node.y)
+
+                            if (i == 0) {
+                                ctx.moveTo(point.x, point.y)
+                            }
+                            else {
+                                ctx.strokeStyle = Theme.rgba(mazeLockSettings.colored ? mazeLock.colors[i] : Theme.primaryColor, 0.5)
+                                ctx.lineTo(point.x, point.y)
+                                ctx.stroke()
+                                ctx.beginPath()
+                                ctx.moveTo(point.x, point.y)
+                            }
+                        }
+                    }
+                }
+            }
+
+            signal activatePosition(int ax, int ay)
+
+            function getNode(index) {
+                var nodeX = parseInt(index % mazeLock.pinx) || mazeLock.pinx
+                var nodeY = parseInt((index - 1) / mazeLock.piny) + 1
+                return Qt.point(nodeX, nodeY)
+            }
+
+            function getNodePoint(px, py) {
+                var posX = pinArea.width / (mazeLock.pinx + 1) * px
+                var posY = pinArea.height / (mazeLock.piny + 1) * py
+                return Qt.point(posX, posY)
+            }
+
+            function processItem(index, nodeX, nodeY) {
+                if (root.enteredPin.length < 2) {
+                    addNode(index, nodeX, nodeY)
+                }
+                else {
+                    addNode(index, nodeX, nodeY)
+                }
+            }
+
+            function addNode(index, nodeX, nodeY) {
+                var nodeChar = mazeLock.chars.charAt(index)
+                if (nodeChar != enteredPin.charAt(enteredPin.length - 1)) {
+                    root._pushPinDigit(nodeChar)
+                    canvas.requestPaint()
+                }
+            }
+
+            function removeNode() {
+                root._popPinDigit()
+                canvas.requestPaint()
+            }
+
+            function cleanNodes() {
+                while (root.enteredPin.length > 0) {
+                    root._popPinDigit()
+                }
+                canvas.requestPaint()
+            }
+
+            onPressed: {
+                cleanNodes()
+            }
+
+            onPressAndHold: {
+                if (root.enteredPin.length == 0) { 
+                    useMaze = false
+                }
+            }
+
+            onPositionChanged: {
+                activatePosition(mouse.x, mouse.y)
+            }
+
+            onReleased: {
+                activatePosition(-1, -1)
+                if (captcha) {
+                    if (root.enteredPin.length >= mazeLock.pinx) {
+                        root._clickedConfirmButton()
+                    }
+                }
+                if (root.enteredPin.length > 4) {
+                    root._clickedConfirmButton()
+                }
+            }
+        }
+
+        Component {
+            id: pinComponent
+            GlassItem {
+                id: nodeItem
+                property bool hovered: false
+                property int activeSize: pinArea.width / (mazeLock.pinx + 1) * (2 / 3)
+                property int size: 64 + (Math.min(4, (root.enteredPin.match(RegExp(mazeLock.chars[index + 1], "g")) || []).length) * 64)
+                width: hovered ? 256 : size
+                height: hovered ? 256 : size
+                property var pinPoint: pinArea.getNode(index + 1)
+                property var nodePoint: pinArea.getNodePoint(pinPoint.x, pinPoint.y)
+                property int posX: nodePoint.x
+                property int posY: nodePoint.y
+                x: posX - width / 2
+                y: posY - height / 2
+                property int thresold: 8
+                z: 10
+
+                Connections {
+                    target: pinArea
+                    onActivatePosition: {
+                        if (!hovered && ax >= posX - (activeSize / 2) + thresold
+                                && ay >= posY - (activeSize / 2) + thresold
+                                && ax <= posX + (activeSize / 2) - thresold
+                                && ay <= posY + (activeSize / 2) - thresold) {
+
+                            pinArea.processItem(index + 1, posX, posY)
+                            hovered = true
+                        }
+                        else if (hovered && (ax < posX - (activeSize / 2)
+                                     || ay < posY - (activeSize / 2)
+                                     || ax > posX + (activeSize / 2)
+                                     || ay > posY + (activeSize / 2))) {
+                            hovered = false
+                        }
+                    }
+                }
+            }
+        }
+    }
+
     PinInputOptionButton {
         id: option1Button
         visible: keypad.visible && text !== "" && showCancelButton
@@ -424,7 +708,7 @@ Item {
         //: Starts the phone call
         //% "Call"
               ? qsTrId("settings_pin-bt-start_call")
-              : (root.enteredPin.length < minimumLength || !root._changedPinValid() ? "" : root.okText)
+              : (useMaze || root.enteredPin.length < minimumLength || !root._changedPinValid() ? "" : root.okText)
         showWhiteBackgroundByDefault: root.emergency
 
         onClicked: {
@@ -442,12 +726,12 @@ Item {
     TextInput {
         id: alphanumProxy
         width: parent.width
-        visible: !keypad.visible
+        visible: !keypad.visible && !useMaze
         y: Qt.inputMethod.keyboardRectangle.y - Theme.itemSizeSmall
 
         horizontalAlignment: TextInput.AlignHCenter
         inputMethodHints: Qt.ImhNoPredictiveText | Qt.ImhSensitiveData | Qt.ImhNoAutoUppercase | Qt.ImhHiddenText
-        focus: !keypad.visible
+        focus: !keypad.visible && !useMaze
         color: root.keypadTextColor
         onCursorVisibleChanged: if (cursorVisible) cursorVisible = false
         font.pixelSize: Theme.fontSizeSmall
@@ -458,6 +742,9 @@ Item {
                     alphanumProxy.forceActiveFocus()
                 }
             }
+            else if (!focus) {
+               root.forceActiveFocus()
+            }
         }
         onTextChanged: {
             if (text.length) root._pushPinDigit(text + "")
@@ -540,13 +827,13 @@ Item {
     }
 
     on_ShowDigitPadChanged: {
-        if (!_showDigitPad) {
+        if (!_showDigitPad && !useMaze) {
             alphanumProxy.forceActiveFocus()
         }
     }
 
     onVisibleInDashboardChanged: {
-        if (!_showDigitPad && visibleInDashboard && enabled) {
+        if (!_showDigitPad && !useMaze && visibleInDashboard && enabled) {
             alphanumProxy.forceActiveFocus()
         }
     }
